### PR TITLE
Tests for trailing commas in function arg lists

### PR DESCRIPTION
**Pull request by jeffmo**
Adds tests for the proposal as described here:
http://jeffmo.github.io/es-trailing-function-commas/

cc @bterlson @ljharb

### CODE REVIEW

**Comment by caitp** [comment]
Looks like there are some good tests missing

- negative tests for `,` following rest parameters
- testing arrow functions, generators, methods, etc

it also might be a bit confusing to mix the Arguments trailing commas and FormalParameters trailing commas tests in the same file, but I don't have that much of an opinion on that

**Comment by leobalter** [comment]
we also need to structure these tests under the existing folders for the functions forms including those listed by @caitp

**Comment by jeffmo** [comment]
> Looks like there are some good tests missing

Good catch! Will add

> we also need to structure these tests under the existing folders for the functions forms

Can you point me to which directories this would be? I wasn't sure if there was a better place for these (or if I named the test file in an acceptable way?)

**Comment by leobalter** [comment]
sure!

- `test/language/expressions/{function,generators,arrow-function}` 
- `test/language/rest-parameters`
- `test/language/statements/{function,generators}` 
- `test/language/expressions/object/method-definition`
- `test/language/statements/class/definition`
- `test/language/{expressions,statements}/async-function`  (see #479)
- `test/built-ins/{Function,GeneratorFunction,AsyncFunction}`

It looks like a lot of tests, but they might be easier if you use the test generation tool available within the repo.

**Comment by tcare** [comment]
Would you mind adding some tests making sure the arguments object is unaffected?

File: test/language/expressions/generators/params-trailing-comma-length.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+// Copyright (C) 2016 Jeff Morrison. All rights reserved.
+// This code is governed by the BSD license found in the LICENSE file.
+/*---
+description: Check that trailing commas are permitted in function argument lists
```
Comment (tcare) [inline test/language/expressions/generators/params-trailing-comma-length.js]: generator function*?

Comment (tcare) [inline test/language/expressions/generators/params-trailing-comma-length.js]: Similar for other descriptions, which should be a bit more descriptive about the individual test's high level purpose

Comment (jeffmo) [inline test/language/expressions/generators/params-trailing-comma-length.js]: Ok. I've updated the generator descriptions to include "generator".

What level of clarity are you looking for for other descriptions though? The current descriptions seem pretty clear to me, but happy to add additional context where necessary.

File: test/language/expressions/arrow-function/params-trailing-comma-length.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+/*---
+description: Check that trailing commas are permitted in function argument lists
+info: http://jeffmo.github.io/es-trailing-function-commas/
+author: Jeff Morrison <lbljeffmo@gmail.com>
+---*/
+
+assert.sameValue(
+  ((a,) => {}).length,
+  1,
+  "Arrow function with 1 arg + trailing comma reports incorrect .length1"
```
Comment (jeffmo) [inline test/language/expressions/arrow-function/params-trailing-comma-length.js]: "length1" (typo -- willfix)

File: test/language/expressions/generators/params-trailing-comma-length.js (hunk 1 above)
Comment (tcare) [inline test/language/expressions/generators/params-trailing-comma-length.js]: Sure, it doesn't have to be as detailed as Mike's tests for example, but something like https://github.com/tc39/test262/pull/735 is a good example.

In this particular case documenting the type of function and/or why the test is interesting (e.g. rest) would be helpful.

**Comment by jeffmo** [comment]
Ok, I've gone through the descriptions and tried to be more specific. Let me know what you think

**Comment by tcare** [comment]
Looks good! Thanks for the update. Would @caitp or @leobalter mind doing a second signoff?

**Comment by caitp** [comment]
Rubberstamp --- I think there are good tests here, which are not tested for all the variants. I am a fan of exhaustive tests, but it is a lot of effort I guess =)

**Comment by tcare** [comment]
I'll merge this for now - @jeffmo feel free to add more tests if you agree with @caitp :)

**Pull request merged**